compiler/luci/pass/src/ForwardReshapeToUnaryOpPass.cpp

```diff
@@ -0,0 +1,188 @@
+ * limitations under the License.
+ */
+
+#include "luci/Pass/ForwardReshapeToUnaryOpPass.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/IR/CircleNodeVisitor.h>
+#include <luci/Log.h>
+#include <luci/Service/CircleShapeInference.h>
+
+namespace
+{
+
+luci::CircleReshape *as_reshape(loco::Node *node)
+{
+  return dynamic_cast<luci::CircleReshape *>(node);
+}
+
+template <loco::DataType T> luci::CircleConst *clone(luci::CircleConst *node)
+{
+  assert(T == node->dtype());
+
+  auto cloned = node->graph()->nodes()->create<luci::CircleConst>();
+  // TODO: We don't have any naming policy for newly created nodes yet.
+  //       Fix this when we have one.
+  cloned->name(node->name());
+  // dtype/shape
+  cloned->dtype(node->dtype());
+  cloned->rank(node->rank());
+  for (uint32_t i = 0; i < node->rank(); i++)
+    cloned->dim(i).set(node->dim(i).value());
+  cloned->shape_status(luci::ShapeStatus::VALID);
+  // values
+  const auto size = node->size<T>();
+  cloned->size<T>(size);
+  for (uint32_t i = 0; i < size; i++)
+    cloned->at<T>(i) = node->at<T>(i);
+  // quantparam
+  const auto *quantparam = node->quantparam();
+  if (quantparam != nullptr)
+  {
+    auto qparam = std::make_unique<luci::CircleQuantParam>();
+    qparam->scale = quantparam->scale;
+    qparam->zerop = quantparam->zerop;
+    qparam->min = quantparam->min;
+    qparam->max = quantparam->max;
+    qparam->quantized_dimension = quantparam->quantized_dimension;
+
+    cloned->quantparam(std::move(qparam));
+  }
+  // sparsity
+  const auto *sparsity = node->sparsityparam();
+  if (sparsity != nullptr)
+  {
+    auto sparam = std::make_unique<luci::SparsityParam>();
+    sparam->traversal_order = sparsity->traversal_order;
+    sparam->block_map = sparsity->block_map;
+    sparam->dim_metadata = sparsity->dim_metadata;
+
+    cloned->sparsityparam(std::move(sparam));
+  }
+  // op version
+  cloned->op_version(node->op_version());
+
+  return cloned;
+}
+
+bool forward_reshape(luci::CircleReshape *reshape, luci::CircleNeg *neg)
+{
+  assert(reshape != nullptr);
+  assert(neg != nullptr);
+
+  luci::CircleConst *cloned_shape = nullptr;
+  const auto reshape_shape = dynamic_cast<luci::CircleConst *>(reshape->shape());
+  // only support CircleConst for now
+  if (reshape_shape == nullptr)
+    return false;
+
+  if (reshape_shape->dtype() == loco::DataType::S32)
+    cloned_shape = clone<loco::DataType::S32>(reshape_shape);
+  else if (reshape_shape->dtype() == loco::DataType::S64)
+    cloned_shape = clone<loco::DataType::S64>(reshape_shape);
+  else
+    return false;
+
+  loco::Graph *graph = neg->graph();
+  // create reshape placed after neg
+  luci::CircleReshape *new_reshape = graph->nodes()->create<luci::CircleReshape>();
+  auto ns_rank = reshape->newShape()->rank();
+  new_reshape->newShape()->rank(ns_rank);
+  for (uint32_t r = 0; r < ns_rank; ++r)
+    new_reshape->newShape()->dim(r) = reshape->newShape()->dim(r);
+
+  new_reshape->shape(cloned_shape);
```

Comment:
Following code is impossible?

```suggestion
  new_reshape->shape(reshape_shape);
```

Reply:
It may work but I wanted to add new items. 
We can later merge same "shape constants" to one and let `Reshape` to use them in another Pass...

Reply:
Then I want to know why creating same constant nodes is better...
If there is no special reason to clone the constant nodes, creating exactly same constant nodes seems redundant for me :(

FYI, merging same constants is being done at export level. :)

Reply:
> Then I want to know why creating same constant nodes is better...

Hum... graph is more simple ?

Reply:
For me, with the reason, benefit of reducing redundant codes(`clone()` function has very long lines) seems larger than the benefit of  creating same constant nodes.
Maybe it is difference of style :-)

Thanks for explanation!